crates/csharp/src/lib.rs

```diff
@@ -1198,50 +1351,195 @@ impl InterfaceGenerator<'_> {
+
+                        [DllImport("{module_name}", EntryPoint = "[resource-drop]{name}"), WasmImportLinkage]
+                        private static extern void wasmImportResourceDrop(int p0);
+        
+                        protected virtual void Dispose(bool disposing) {{
+                            if (Handle.HasValue) {{
+                                wasmImportResourceDrop((int) Handle);
+                                Handle = null;
+                            }}
+                        }}
+                    "#
+                );
             }
-            count => {
-                self.gen.tuple_counts.insert(count);
-                format!(
-                    "({})",
-                    func.results
-                        .iter_types()
-                        .map(|ty| self.type_name_boxed(ty, qualifier))
-                        .collect::<Vec<_>>()
-                        .join(", ")
-                )
+            Direction::Export => {
+                let prefix = key
+                    .map(|s| format!("{}#", self.resolve.name_world_key(s)))
+                    .unwrap_or_else(String::new);
+
+                uwrite!(
+                    self.csharp_interop_src,
+                    r#"
+                    [UnmanagedCallersOnly(EntryPoint = "{prefix}[dtor]{name}")]
+                    public static unsafe void wasmExportResourceDtor{upper_camel}(int rep) {{
+                        var val = ({qualified}) RepTable.Remove(rep);
+                        val.Dispose();
+                    }}
+                    "#
+                );
+
+                let module_name = key
+                    .map(|key| format!("[export]{}", self.resolve.name_world_key(key)))
+                    .unwrap_or_else(|| "[export]$root".into());
+
+                uwriteln!(
+                    self.src,
+                    r#"
+                    public abstract class {upper_camel}: IDisposable {{
+                        internal int? Handle {{ get; set; }}
+
+                        public void Dispose() {{
+                            Dispose(true);
+                            GC.SuppressFinalize(this);
+                        }}
+
+                        internal static class WasmInterop {{
+                            [DllImport("{module_name}", EntryPoint = "[resource-drop]{name}"), WasmImportLinkage]
+                            internal static extern void wasmImportResourceDrop(int p0);
+
+                            [DllImport("{module_name}", EntryPoint = "[resource-new]{name}"), WasmImportLinkage]
+                            internal static extern int wasmImportResourceNew(int p0);
+                    
+                            [DllImport("{module_name}", EntryPoint = "[resource-rep]{name}"), WasmImportLinkage]
+                            internal static extern int wasmImportResourceRep(int p0);
+                        }}
+
+                        protected virtual void Dispose(bool disposing) {{
```

Comment:
Is there a value in `IDisposable` for the export.  There isn't anything unmanaged to worry about, I'd think we could just drop it from `RepTable`

Reply:
I _think_ we could technically get rid of it, but there's an argument for keeping it:

When a component passes ownership of an exported resource instance to the host, it calls the `[resource-new]` function to allocate a handle for it.  Later, if the host passes ownership of that same instance back to the component, the component could technically dispose of the handle immediately using `[resource-drop]`, but if it wants to pass ownership back to the host again, it will have to call `[resource-new]` and allocate a new handle.  By keeping the original handle alive, we avoid the churn of calling `[resource-new]` and `[resource-drop]` repeatedly, which could noticeably reduce overhead in a tight loop.  The drawback, as you point out, is that we need to implement `IDisposable` to ensure we call `[resource-drop]` once the instance is no longer needed.

Reply:
Well now that I look at the code I wrote, it doesn't seem to be tracking the handle the way I described above, so I'll dig into it some more.

Reply:
Okay, _now_ the code does what I described above.

Reply:
I've added a comment.

Reply:
Thanks, I didn't appreciate that the ownership could be passed over.